### PR TITLE
API Delete deprecated canArchive method

### PR DESCRIPTION
**Pull request by GuySartorelli**
For versioned records, there is no concept of "deleting" the record - it's either being archived, unpublished, or in rare cases removed from draft but left published.

Previously the `canDelete()` method has been used to check if a record can be removed from draft specifically for versioned records - which is such a rare edge case it almost doesn't bare thinking about, especially given there's no way to do that purely by interacting with the UI, nor with the web API endpoints available in core or supported modules.

Instead, `canDelete()` is now to be used to check if the record can be archived. This reduces cognitive load for developers and is a step towards simplifying the overly-complex versioning system. This is also in keeping with other mechanisms, such as `cascade_deletes` which cascades archiving for versioned records.

Note that if the record has stages and is published, we don't allow users to archive if they can't unpublish.

## Issue
- https://github.com/silverstripe/silverstripe-versioned/issues/447

### CODE REVIEW

File: src/GridFieldArchiveAction.php [hunk 1]
```diff
@@ -190,13 +191,13 @@ public function handleAction(GridField $gridField, $actionName, $arguments, $dat
...
-     * @param DataObject $record
+     * @param ViewableData $record
      * @return GridField_FormAction|null
      */
     public function getArchiveAction($gridField, $record)
     {
-        /* @var DataObject|Versioned $record */
-        if (!$record->hasMethod('canArchive') || !$record->canArchive()) {
+        /** @var ViewableData|Versioned $record */
+        if (!$record->has_extension(Versioned::class) || !$record->canDelete()) {
```
Comment (ChloeSartorelli) [inline src/GridFieldArchiveAction.php]: Note that the method check used to be there to distinguish between versioned and unversioned records. `has_extension()` is the correct way to do that, so we're doing it that way instead.

File: src/GridFieldArchiveAction.php [hunk 2]
```diff
@@ -190,13 +191,13 @@ public function handleAction(GridField $gridField, $actionName, $arguments, $dat
      * Returns the GridField_FormAction if archive can be performed
      *
      * @param GridField $gridField
-     * @param DataObject $record
+     * @param ViewableData $record
```
Comment (ChloeSartorelli) [inline src/GridFieldArchiveAction.php:194]: semi-unrelated PHPDoc fix - gridfield can accept any viewable data. This class is used to replace the delete button when versioned is installed, regardless of whether the gridfield in use has dataobjects or not.

**Review by emteknetnz — APPROVED** [review]
